File: src/main/java/com/example/the_machine/domain/MessageEntity.java
package com.example.the_machine.domain;

import com.fasterxml.jackson.databind.JsonNode;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "message")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class MessageEntity {

  @Id
  private UUID id;

  @Column(nullable = false)
  private UUID threadId;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private Author author;

  @Column(nullable = false)
  private Instant createdAt;

  @Column(nullable = false, columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  private JsonNode contentJson;

  @Column(columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  private JsonNode attachmentsJson;

  public enum Author {
    USER, ASSISTANT, TOOL, SYSTEM
  }
}

File: src/main/kotlin/com/example/the_machine/domain/MessageEntity.kt
package com.example.the_machine.domain

import com.example.the_machine.common.JsonElementConverter
import com.example.the_machine.common.KotlinSerializationConfig
import jakarta.persistence.*
import kotlinx.serialization.json.JsonElement
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes
import java.util.*

@Entity
@Table(name = "message")
data class MessageEntity(
  @Column(nullable = false)
  val threadId: UUID,

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  val author: Author,

  @Column(nullable = false, columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  @Convert(converter = JsonElementConverter::class)
  val contentJson: JsonElement = KotlinSerializationConfig.staticJson.parseToJsonElement("[]"),

  @Column(columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  @Convert(converter = JsonElementConverter::class)
  val attachmentsJson: JsonElement? = null
) : BaseEntity() {

  enum class Author {
    USER, ASSISTANT, TOOL, SYSTEM
  }

}